--- compiled-release-x86_64/dist/nightingale.orig	2011-09-26 22:59:56.000000000 -0500
+++ compiled-release-x86_64/dist/nightingale	2011-09-26 22:59:46.000000000 -0500
@@ -45,6 +45,19 @@
 MOZ_VIEWER_NAME="./viewer"
 MOZ_PROGRAM=""
 
+# this depends on your system's gstreamer location
+# while sometimes, the make process actually uses the location properly
+# it doesn't always, and this is better than symlinking
+for dir in /usr/lib64 /usr/lib ; do
+  if [ -f ${dir}/gstreamer-0.10/libgstcoreelements.so ] ; then
+    export GST_PLUGIN_PATH=${dir}/gstreamer\-0.10
+    break
+  elif [ -f ${dir}/gstreamer0.10/libgstcoreelements.so ] ; then
+	export GST_PLUIN_PATH=${dir}/gstreamer0.10
+	break
+  fi
+done
+
 exitcode=0
 
 #
