Index: pixman-remap.h
===================================================================
RCS file: /cvsroot/mozilla/gfx/cairo/libpixman/src/pixman-remap.h,v
retrieving revision 1.1.6.2
diff -u -p -r1.1.6.2 pixman-remap.h
--- pixman-remap.h	27 Sep 2005 23:09:03 -0000	1.1.6.2
+++ pixman-remap.h	21 Nov 2005 02:35:24 -0000
@@ -1,9 +1,7 @@
 #define pixman_add_trapezoids _cairo_pixman_add_trapezoids
-#define INT_pixman_color_to_pixel _cairo_pixman_color_to_pixel
 #define pixman_color_to_pixel _cairo_pixman_color_to_pixel
 #define composeFunctions _cairo_pixman_compose_functions
 #define fbComposeSetupMMX _cairo_pixman_compose_setup_mmx
-#define INT_pixman_composite _cairo_pixman_composite
 #define pixman_composite _cairo_pixman_composite
 #define fbCompositeCopyAreammx _cairo_pixman_composite_copy_area_mmx
 #define fbCompositeSolidMask_nx8888x0565Cmmx _cairo_pixman_composite_solid_mask_nx8888x0565Cmmx
