Index: xpcom/glue/nsTArray.cpp
===================================================================
RCS file: /cvsroot/mozilla/xpcom/glue/nsTArray.cpp,v
retrieving revision 1.7
diff -u -p -r1.7 nsTArray.cpp
--- xpcom/glue/nsTArray.cpp	8 Jul 2007 07:08:51 -0000	1.7
+++ xpcom/glue/nsTArray.cpp	10 Jul 2007 17:29:34 -0000
@@ -69,7 +69,10 @@ nsTArray_base::EnsureCapacity(size_type 
     return PR_FALSE;
   }
 
-  if (mHdr == &sEmptyHdr) {
+  // XXX Somehow mHdr doesn't always point to the right sEmptyHdr (linking
+  //     problems on OS X in debug builds). We're going to instead check the
+  //     capacity of the header.
+  if (!mHdr->mCapacity) {
     // NS_Alloc new data
     Header *header = static_cast<Header*>
                                 (NS_Alloc(sizeof(Header) + capacity * elemSize));
