Index: extensions/inspector/install.rdf
===================================================================
RCS file: /cvsroot/mozilla/extensions/inspector/install.rdf,v
retrieving revision 1.14
diff -u -8 -p -r1.14 install.rdf
--- extensions/inspector/install.rdf	10 Jun 2007 19:30:14 -0000	1.14
+++ extensions/inspector/install.rdf	28 Jun 2007 22:14:43 -0000
@@ -39,16 +39,25 @@
       <!-- Suite -->
       <Description>
         <em:id>{92650c4d-4b8e-4d2a-b7eb-24ecf4f6b63a}</em:id>
         <em:minVersion>1.5a</em:minVersion>
         <em:maxVersion>@SEAMONKEY_VERSION@</em:maxVersion>
       </Description>
     </em:targetApplication>
     
+    <em:targetApplication>
+      <!-- Songbird -->
+      <Description>
+        <em:id>songbird@songbirdnest.com</em:id>
+        <em:minVersion>0.2</em:minVersion>
+        <em:maxVersion>0.5</em:maxVersion>
+      </Description>
+    </em:targetApplication>
+    
     <!-- EXTENSION AUTHORS!
          DO NOT COPY THIS PROPERTY INTO YOUR INSTALL RDF FILES
          It will cause users not to be informed of incompatibilities 
          with your extension when they are updated with Software Update
          and your extension will become unavailable to them!
          -->
     <em:appManaged>true</em:appManaged>
 
Index: extensions/inspector/jar.mn
===================================================================
RCS file: /cvsroot/mozilla/extensions/inspector/jar.mn,v
retrieving revision 1.29
diff -u -8 -p -r1.29 jar.mn
--- extensions/inspector/jar.mn	17 Jun 2007 04:22:49 -0000	1.29
+++ extensions/inspector/jar.mn	28 Jun 2007 22:14:43 -0000
@@ -11,16 +11,17 @@ inspector.jar:
 % overlay chrome://messenger/content/mailWindowOverlay.xul chrome://inspector/content/tasksOverlay-tb.xul application={3550f703-e582-4d05-9a08-453d09bdfdc6}
 % overlay chrome://inspector/content/commandOverlay.xul chrome://inspector/content/viewers/dom/commandOverlay.xul
 % overlay chrome://inspector/content/commandOverlay.xul chrome://inspector/content/viewers/styleRules/commandOverlay.xul
 % overlay chrome://inspector/content/commandOverlay.xul chrome://inspector/content/search/modules/commandOverlay.xul
 % overlay chrome://inspector/content/keysetOverlay.xul chrome://inspector/content/viewers/dom/keysetOverlay.xul
 % overlay chrome://inspector/content/popupOverlay.xul chrome://inspector/content/viewers/dom/popupOverlay.xul
 % overlay chrome://inspector/content/popupOverlay.xul chrome://inspector/content/viewers/styleRules/popupOverlay.xul
 % overlay chrome://inspector/content/popupOverlay.xul chrome://inspector/content/search/modules/popupOverlay.xul
+% overlay chrome://songbird/content/xul/menuOverlay.xul chrome://inspector/content/tasksOverlay-songbird.xul application=songbird@songbirdnest.com
   content/inspector/inspector.xul                                     (resources/content/inspector.xul)
   content/inspector/inspector.js                                      (resources/content/inspector.js)
   content/inspector/inspector.css                                     (resources/content/inspector.css)
   content/inspector/inspector.xml                                     (resources/content/inspector.xml)
   content/inspector/sidebar.xul                                       (resources/content/sidebar.xul)
   content/inspector/sidebar.js                                        (resources/content/sidebar.js)
   content/inspector/object.xul                                        (resources/content/object.xul)
   content/inspector/object.js                                         (resources/content/object.js)
@@ -28,16 +29,17 @@ inspector.jar:
   content/inspector/toolboxOverlay.xul                                (resources/content/toolboxOverlay.xul)
   content/inspector/commandOverlay.xul                                (resources/content/commandOverlay.xul)
   content/inspector/keysetOverlay.xul                                 (resources/content/keysetOverlay.xul)
   content/inspector/popupOverlay.xul                                  (resources/content/popupOverlay.xul)
   content/inspector/statusbarOverlay.xul                              (resources/content/statusbarOverlay.xul)
   content/inspector/tasksOverlay.xul                                  (resources/content/tasksOverlay.xul)
   content/inspector/tasksOverlay-ff.xul                               (resources/content/tasksOverlay-ff.xul)
   content/inspector/tasksOverlay-sb.xul                               (resources/content/tasksOverlay-sb.xul)
+  content/inspector/tasksOverlay-songbird.xul                         (resources/content/tasksOverlay-songbird.xul)
   content/inspector/tasksOverlay-tb.xul                               (resources/content/tasksOverlay-tb.xul)
   content/inspector/editingOverlay.xul                                (resources/content/editingOverlay.xul)
   content/inspector/utilWindow.xul                                    (resources/content/utilWindow.xul)
   content/inspector/Flasher.js                                        (resources/content/Flasher.js)
   content/inspector/ViewerRegistry.js                                 (resources/content/ViewerRegistry.js)
   content/inspector/utils.js                                          (resources/content/utils.js)
   content/inspector/hooks.js                                          (resources/content/hooks.js)
   content/inspector/util.dtd                                          (resources/content/util.dtd)
Index: extensions/inspector/resources/content/tasksOverlay-songbird.xul
===================================================================
RCS file: extensions/inspector/resources/content/tasksOverlay-songbird.xul
diff -N extensions/inspector/resources/content/tasksOverlay-songbird.xul
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ extensions/inspector/resources/content/tasksOverlay-songbird.xul	28 Jun 2007 22:14:43 -0000
@@ -0,0 +1,43 @@
+<?xml version="1.0"?>
+
+<!DOCTYPE overlay SYSTEM "chrome://inspector/locale/tasksOverlay.dtd" >
+
+<overlay id="eInspectorTasksOverlay"
+         xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul">
+
+  <script type="application/x-javascript" src="chrome://inspector/content/hooks.js"/>
+  <script type="application/x-javascript">
+  <![CDATA[
+    function inspectorWindowLoad()
+    {
+      setTimeout(initWindowMediator, 0);
+    }
+
+    function initWindowMediator()
+    {
+      // Load the Window DataSource so that browser windows opened subsequent to DOM
+      // Inspector show up in the DOM Inspector's window list.
+      var windowDS = Components.classes["@mozilla.org/rdf/datasource;1?name=window-mediator"]
+                               .getService(Components.interfaces.nsIWindowDataSource);
+    }
+
+    window.addEventListener("load", inspectorWindowLoad, false);
+  ]]>
+  </script>
+
+  <commandset id="mainCommandSet">
+    <command id="Tasks:Inspector" oncommand="inspectDOMDocument();"/>
+    <command id="Tasks:InspectPage" oncommand="inspectDOMDocument(content.document);"/>
+  </commandset>
+
+  <keyset id="mainKeyset">
+    <key id="key_inspectPage" key="&inspectorTaskCmd.commandkey;" modifiers="accel,shift" command="Tasks:Inspector"/>
+  </keyset>
+  
+  <menupopup id="menu_ToolsPopup">
+    <menuitem id="menu_inspector" label="&inspectorTaskCmd.label;"
+              insertafter="javascriptConsole" key="key_inspectPage"
+              accesskey="&inspectorTaskCmd.accesskey.ff;" command="Tasks:Inspector"/>
+  </menupopup>
+
+</overlay>
