Index: xulrunner/stub/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/xulrunner/stub/Makefile.in,v
retrieving revision 1.7
diff -u -p -r1.7 Makefile.in
--- xulrunner/stub/Makefile.in	31 Jan 2007 20:29:11 -0000	1.7
+++ xulrunner/stub/Makefile.in	27 Sep 2007 03:43:16 -0000
@@ -51,15 +51,7 @@ CPPSRCS   = nsXULStub.cpp
 # Statically link the CRT when possible
 USE_STATIC_LIBS = 1
 
-ifeq ($(OS_ARCH),Darwin)
-
-PROGRAM   = xulrunner$(BIN_SUFFIX)
-
-else
-
-PROGRAM   = xulrunner-stub$(BIN_SUFFIX)
-
-endif # OS_ARCH
+PROGRAM   = Songbird$(BIN_SUFFIX)
 
 LOCAL_INCLUDES = -I$(topsrcdir)/xpcom/build
 
