? bug392526-nsID-toString-mismatched-allocator.patch
Index: xpcom/glue/nsID.cpp
===================================================================
RCS file: /cvsroot/mozilla/xpcom/glue/nsID.cpp,v
retrieving revision 3.12
diff -p -u -8 -r3.12 nsID.cpp
--- xpcom/glue/nsID.cpp	9 Aug 2005 14:04:49 -0000	3.12
+++ xpcom/glue/nsID.cpp	20 Oct 2007 00:26:59 -0000
@@ -33,16 +33,17 @@
  * and other provisions required by the GPL or the LGPL. If you do not delete
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 #include "nsID.h"
 #include "prprf.h"
 #include "prmem.h"
+#include "nsXPCOM.h"
 
 static const char gIDFormat[] = 
   "{%08x-%04x-%04x-%02x%02x-%02x%02x%02x%02x%02x%02x}";
 
 static const char gIDFormat2[] = 
   "%08x-%04x-%04x-%02x%02x-%02x%02x%02x%02x%02x%02x";
 
 
@@ -121,17 +122,17 @@ PRBool nsID::Parse(const char *aIDStr)
 /*
  * Returns an allocated string in {xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}
  * format. The string is allocated with PR_Malloc and should be freed by
  * the caller.
  */
 
 char *nsID::ToString() const 
 {
-  char *res = (char*)PR_Malloc(39);    // use PR_Malloc if this is to be freed with nsCRT::free
+  char *res = (char*)NS_Alloc(39);    // may be passed across XPCOM boundaries, use NS_Alloc
 
   if (res != NULL) {
     PR_snprintf(res, 39, gIDFormat,
                 m0, (PRUint32) m1, (PRUint32) m2,
                 (PRUint32) m3[0], (PRUint32) m3[1], (PRUint32) m3[2],
                 (PRUint32) m3[3], (PRUint32) m3[4], (PRUint32) m3[5],
                 (PRUint32) m3[6], (PRUint32) m3[7]);
   }
Index: xpcom/reflect/xptinfo/src/xptiManifest.cpp
===================================================================
RCS file: /cvsroot/mozilla/xpcom/reflect/xptinfo/src/xptiManifest.cpp,v
retrieving revision 1.35
diff -p -u -8 -r1.35 xptiManifest.cpp
--- xpcom/reflect/xptinfo/src/xptiManifest.cpp	19 Oct 2007 23:26:53 -0000	1.35
+++ xpcom/reflect/xptinfo/src/xptiManifest.cpp	20 Oct 2007 00:26:59 -0000
@@ -106,17 +106,17 @@ xpti_InterfaceWriter(PLDHashTable *table
                                    (int) number,
                                    entry->GetTheName(),
                                    iidStr,
                                    (int) typelib.GetFileIndex(),
                                    (int) (typelib.IsZip() ? 
                                    typelib.GetZipItemIndex() : -1),
                                    (int) entry->GetScriptableFlag());
 
-    nsCRT::free(iidStr);
+    NS_Free(iidStr);
 
     return success ? PL_DHASH_NEXT : PL_DHASH_STOP;
 }
 
 
 // static
 PRBool xptiManifest::Write(xptiInterfaceInfoManager* aMgr,
                            xptiWorkingSet*           aWorkingSet)
