Index: js/src/xpconnect/src/XPCWrapper.cpp
===================================================================
RCS file: /cvsroot/mozilla/js/src/xpconnect/src/XPCWrapper.cpp,v
retrieving revision 1.6
diff -u -p -r1.6 XPCWrapper.cpp
--- js/src/xpconnect/src/XPCWrapper.cpp	1 Oct 2007 00:31:22 -0000	1.6
+++ js/src/xpconnect/src/XPCWrapper.cpp	9 Oct 2007 01:21:42 -0000
@@ -167,12 +167,16 @@ XPCWrapper::NewResolve(JSContext *cx, JS
   }
 
   JSBool isXOW = (JS_GET_CLASS(cx, wrapperObj) == &sXPC_XOW_JSClass.base);
-  JSScopeProperty *sprop = reinterpret_cast<JSScopeProperty *>(prop);
-  uintN attrs = sprop->attrs;
-  if ((preserveVal || isXOW) &&
-      SPROP_HAS_VALID_SLOT(sprop, OBJ_SCOPE(innerObjp))) {
-    v = OBJ_GET_SLOT(cx, innerObjp, sprop->slot);
-  }
+  uintN attrs = JSPROP_ENUMERATE;
+  if (OBJ_IS_NATIVE(innerObjp)) {
+    JSScopeProperty *sprop = reinterpret_cast<JSScopeProperty *>(prop);
+    attrs = sprop->attrs;
+    if ((preserveVal || isXOW) &&
+        SPROP_HAS_VALID_SLOT(sprop, OBJ_SCOPE(innerObjp))) {
+      v = OBJ_GET_SLOT(cx, innerObjp, sprop->slot);
+    }
+  }
+
   OBJ_DROP_PROPERTY(cx, innerObjp, prop);
 
   // Hack alert: we only do this for same-origin calls on XOWs: we want
