Index: toolkit/xre/nsAppRunner.cpp
===================================================================
RCS file: /cvsroot/mozilla/toolkit/xre/nsAppRunner.cpp,v
retrieving revision 1.192
diff -u -8 -p -r1.192 nsAppRunner.cpp
--- toolkit/xre/nsAppRunner.cpp	20 Sep 2007 16:54:00 -0000	1.192
+++ toolkit/xre/nsAppRunner.cpp	26 Sep 2007 21:13:49 -0000
@@ -2561,19 +2561,18 @@ XRE_main(int argc, char* argv[], const n
     appData.flags |= NS_XRE_ENABLE_CRASH_REPORTER;
   }
 
   if ((appData.flags & NS_XRE_ENABLE_CRASH_REPORTER) &&
       NS_SUCCEEDED(
          CrashReporter::SetExceptionHandler(appData.xreDirectory,
                                             appData.crashReporterURL))) {
     // pass some basic info from the app data
-    if (appData.vendor)
-      CrashReporter::AnnotateCrashReport(NS_LITERAL_CSTRING("Vendor"),
-                                     nsDependentCString(appData.vendor));
+    CrashReporter::AnnotateCrashReport(NS_LITERAL_CSTRING("Vendor"),
+                                       NS_LITERAL_CSTRING("Songbird"));
     if (appData.name)
       CrashReporter::AnnotateCrashReport(NS_LITERAL_CSTRING("ProductName"),
                                      nsDependentCString(appData.name));
     if (appData.version)
       CrashReporter::AnnotateCrashReport(NS_LITERAL_CSTRING("Version"),
                                      nsDependentCString(appData.version));
     if (appData.buildID)
       CrashReporter::AnnotateCrashReport(NS_LITERAL_CSTRING("BuildID"),
