Index: xpcom/glue/nsTArray.h
===================================================================
RCS file: /cvsroot/mozilla/xpcom/glue/nsTArray.h,v
retrieving revision 1.33
diff -u -8 -p -r1.33 nsTArray.h
--- xpcom/glue/nsTArray.h	26 Dec 2006 14:24:15 -0000	1.33
+++ xpcom/glue/nsTArray.h	1 May 2007 19:40:58 -0000
@@ -165,17 +165,17 @@ class NS_COM_GLUE nsTArray_base {
     // built-in buffer to store its elements.
     PRBool UsesAutoArrayBuffer() {
       return mHdr->mIsAutoArray && mHdr == GetAutoArrayBuffer();
     }
 
     // This is not const since we may actually write to it. However we will
     // always write to it the same data that it already contains. See
     // IncrementLength
-    static Header sEmptyHdr;
+    static NS_HIDDEN_(Header) sEmptyHdr;
 
     // The array's elements (prefixed with a Header).  This pointer is never
     // null.  If the array is empty, then this will point to sEmptyHdr.
     Header *mHdr;
 };
 
 //
 // This class defines convenience functions for element specific operations.
