Index: xpcom/glue/nsTArray.cpp
===================================================================
RCS file: /cvsroot/mozilla/xpcom/glue/nsTArray.cpp,v
retrieving revision 1.16
diff -u -p -r1.16 nsTArray.cpp
--- xpcom/glue/nsTArray.cpp	25 Jan 2008 11:32:59 -0000	1.16
+++ xpcom/glue/nsTArray.cpp	1 Feb 2008 18:40:28 -0000
@@ -70,7 +70,10 @@ nsTArray_base::EnsureCapacity(size_type 
     return PR_FALSE;
   }
 
-  if (mHdr == &sEmptyHdr) {
+  // XXX Somehow mHdr doesn't always point to the right sEmptyHdr (linking
+  //     problems on OS X in debug builds). We're going to instead check the
+  //     capacity of the header.
+  if (!mHdr->mCapacity) {
     // NS_Alloc new data
     Header *header = static_cast<Header*>
                                 (NS_Alloc(sizeof(Header) + capacity * elemSize));
