Index: widget/src/cocoa/nsNativeThemeCocoa.h
===================================================================
RCS file: /cvsroot/mozilla/widget/src/cocoa/nsNativeThemeCocoa.h,v
retrieving revision 1.20
diff -u -8 -r1.20 nsNativeThemeCocoa.h
--- widget/src/cocoa/nsNativeThemeCocoa.h	9 May 2008 18:12:39 -0000	1.20
+++ widget/src/cocoa/nsNativeThemeCocoa.h	30 Jun 2008 18:58:19 -0000
@@ -95,17 +95,17 @@
   nsresult GetSystemFont(PRUint8 aWidgetType, nsSystemFontID& aFont);
 
   // HITheme drawing routines
   void DrawFrame (CGContextRef context, HIThemeFrameKind inKind,
                   const HIRect& inBoxRect, PRBool inIsDisabled,
                   PRInt32 inState);
   void DrawProgress(CGContextRef context, const HIRect& inBoxRect,
                     PRBool inIsIndeterminate, PRBool inIsHorizontal,
-                    PRInt32 inValue);
+                    PRInt32 inValue, PRInt32 inMaxValue);
   void DrawTab (CGContextRef context, const HIRect& inBoxRect,
                 PRBool inIsDisabled, PRBool inIsFrontmost, 
                 PRBool inIsHorizontal, PRBool inTabBottom,
                 PRInt32 inState);
   void DrawTabPanel (CGContextRef context, const HIRect& inBoxRect);
   void DrawScale (CGContextRef context, const HIRect& inBoxRect,
                   PRBool inIsDisabled, PRInt32 inState,
                   PRBool inDirection, PRBool inIsReverse,
Index: widget/src/cocoa/nsNativeThemeCocoa.mm
===================================================================
RCS file: /cvsroot/mozilla/widget/src/cocoa/nsNativeThemeCocoa.mm,v
retrieving revision 1.93
diff -u -8 -r1.93 nsNativeThemeCocoa.mm
--- widget/src/cocoa/nsNativeThemeCocoa.mm	9 May 2008 18:12:39 -0000	1.93
+++ widget/src/cocoa/nsNativeThemeCocoa.mm	30 Jun 2008 18:58:19 -0000
@@ -667,31 +667,34 @@
   HIThemeDrawFrame(&drawRect, &fdi, cgContext, HITHEME_ORIENTATION);
 
   NS_OBJC_END_TRY_ABORT_BLOCK;
 }
 
 
 void
 nsNativeThemeCocoa::DrawProgress(CGContextRef cgContext,
-                                 const HIRect& inBoxRect, PRBool inIsIndeterminate, 
-                                 PRBool inIsHorizontal, PRInt32 inValue)
+                                 const HIRect& inBoxRect, 
+                                 PRBool inIsIndeterminate, 
+                                 PRBool inIsHorizontal, 
+                                 PRInt32 inValue, 
+                                 PRInt32 inMaxValue)
 {
   NS_OBJC_BEGIN_TRY_ABORT_BLOCK;
 
   HIThemeTrackDrawInfo tdi;
 
   PRInt32 stepsPerSecond = inIsIndeterminate ? 60 : 30;
   PRInt32 milliSecondsPerStep = 1000 / stepsPerSecond;
 
   tdi.version = 0;
   tdi.kind = inIsIndeterminate ? kThemeMediumIndeterminateBar: kThemeMediumProgressBar;
   tdi.bounds = inBoxRect;
   tdi.min = 0;
-  tdi.max = 100;
+  tdi.max = inMaxValue;
   tdi.value = inValue;
   tdi.attributes = inIsHorizontal ? kThemeTrackHorizontal : 0;
   tdi.enableState = kThemeTrackActive;
   tdi.trackInfo.progress.phase = PR_IntervalToMilliseconds(PR_IntervalNow()) /
                                  milliSecondsPerStep % 16;
 
   HIThemeDrawTrack(&tdi, NULL, cgContext, HITHEME_ORIENTATION);
 
@@ -1190,22 +1193,22 @@
       }
 
       DrawFrame(cgContext, kHIThemeFrameTextFieldSquare,
                 macRect, (IsDisabled(aFrame) || IsReadOnly(aFrame)), eventState);
       break;
       
     case NS_THEME_PROGRESSBAR:
       DrawProgress(cgContext, macRect, IsIndeterminateProgress(aFrame),
-                   PR_TRUE, GetProgressValue(aFrame));
+                   PR_TRUE, GetProgressValue(aFrame), GetProgressMaxValue(aFrame));
       break;
 
     case NS_THEME_PROGRESSBAR_VERTICAL:
       DrawProgress(cgContext, macRect, IsIndeterminateProgress(aFrame),
-                   PR_FALSE, GetProgressValue(aFrame));
+                   PR_FALSE, GetProgressValue(aFrame), GetProgressMaxValue(aFrame));
       break;
 
     case NS_THEME_PROGRESSBAR_CHUNK:
     case NS_THEME_PROGRESSBAR_CHUNK_VERTICAL:
       // do nothing, covered by the progress bar cases above
       break;
 
     case NS_THEME_TREEVIEW_TWISTY:
Index: widget/src/xpwidgets/nsNativeTheme.h
===================================================================
RCS file: /cvsroot/mozilla/widget/src/xpwidgets/nsNativeTheme.h,v
retrieving revision 1.28
diff -u -8 -r1.28 nsNativeTheme.h
--- widget/src/xpwidgets/nsNativeTheme.h	27 Mar 2008 05:36:21 -0000	1.28
+++ widget/src/xpwidgets/nsNativeTheme.h	30 Jun 2008 18:58:20 -0000
@@ -128,16 +128,20 @@
   PRBool IsHorizontal(nsIFrame* aFrame);
 
   // progressbar:
   PRBool IsIndeterminateProgress(nsIFrame* aFrame);
 
   PRInt32 GetProgressValue(nsIFrame* aFrame) {
     return CheckIntAttr(aFrame, nsWidgetAtoms::value, 0);
   }
+  
+  PRInt32 GetProgressMaxValue(nsIFrame* aFrame) {
+    return PR_MAX(CheckIntAttr(aFrame, nsWidgetAtoms::max, 100), 1);
+  }
 
   // textfield:
   PRBool IsReadOnly(nsIFrame* aFrame) {
       return CheckBooleanAttr(aFrame, nsWidgetAtoms::readonly);
   }
 
   nsIPresShell *GetPresShell(nsIFrame* aFrame);
   PRInt32 CheckIntAttr(nsIFrame* aFrame, nsIAtom* aAtom, PRInt32 defaultValue);
Index: widget/src/xpwidgets/nsWidgetAtomList.h
===================================================================
RCS file: /cvsroot/mozilla/widget/src/xpwidgets/nsWidgetAtomList.h,v
retrieving revision 1.26
diff -u -8 -r1.26 nsWidgetAtomList.h
--- widget/src/xpwidgets/nsWidgetAtomList.h	7 Apr 2008 22:56:44 -0000	1.26
+++ widget/src/xpwidgets/nsWidgetAtomList.h	30 Jun 2008 18:58:20 -0000
@@ -80,16 +80,17 @@
 WIDGET_ATOM(horizontal, "horizontal")
 WIDGET_ATOM(vertical, "vertical")
 WIDGET_ATOM(id, "id")
 WIDGET_ATOM(image, "image")
 WIDGET_ATOM(input, "input")
 WIDGET_ATOM(key, "key") // The key element / attribute
 WIDGET_ATOM(label, "label")
 WIDGET_ATOM(lasttab, "last-tab")
+WIDGET_ATOM(max, "max")
 WIDGET_ATOM(maxpos, "maxpos")
 WIDGET_ATOM(menu, "menu") // Represents an XP menu
 WIDGET_ATOM(menuitem, "menuitem") // Represents an XP menu item
 WIDGET_ATOM(menupopup, "menupopup") // The XP menu's children.
 WIDGET_ATOM(menuseparator, "menuseparator")  // Divider between menu items
 WIDGET_ATOM(menuFrame, "MenuFrame")
 WIDGET_ATOM(minpos, "minpos")
 WIDGET_ATOM(mode, "mode")
