Index: toolkit/mozapps/downloads/content/helperApps.js
===================================================================
RCS file: /cvsroot/mozilla/toolkit/mozapps/downloads/content/helperApps.js,v
retrieving revision 1.12
diff -u -8 -p -r1.12 helperApps.js
--- toolkit/mozapps/downloads/content/helperApps.js	5 Jul 2007 19:31:45 -0000	1.12
+++ toolkit/mozapps/downloads/content/helperApps.js	5 Oct 2007 16:17:16 -0000
@@ -620,17 +620,25 @@ HandlerOverride.prototype = {
     var helperAppResource = gRDF.GetUnicodeResource(APP_URI(this.mimeType));
     this._DS.Assert(handlerResource, helperAppProperty, helperAppResource, true);
     // add the mime type to the MIME types seq
     var container = Components.classes["@mozilla.org/rdf/container;1"].createInstance();
     if (container) {
       container = container.QueryInterface(Components.interfaces.nsIRDFContainer);
       if (container) {
         var containerRes = gRDF.GetUnicodeResource("urn:mimetypes:root");
-        container.Init(this._DS, containerRes);
+        try {
+          container.Init(this._DS, containerRes);
+        } catch (e) {
+          var containerUtils =
+              Components.classes["@mozilla.org/rdf/container-utils;1"]
+                        .getService(Components.interfaces.nsIRDFContainerUtils);
+          containerUtils.MakeSeq(this._DS, containerRes);
+          container.Init(this._DS, containerRes);
+        }
         var element = gRDF.GetUnicodeResource(MIME_URI(this.mimeType));
         if (container.IndexOf(element) == -1)
           container.AppendElement(element);
       }
     }
   }, 
   
   // Implementation helper methods
