Index: layout/xul/base/src/tree/src/nsTreeBodyFrame.cpp
===================================================================
RCS file: /cvsroot/mozilla/layout/xul/base/src/tree/src/nsTreeBodyFrame.cpp,v
retrieving revision 1.335
diff -u -p -r1.335 nsTreeBodyFrame.cpp
--- layout/xul/base/src/tree/src/nsTreeBodyFrame.cpp	15 Oct 2007 11:21:25 -0000	1.335
+++ layout/xul/base/src/tree/src/nsTreeBodyFrame.cpp	26 Oct 2007 03:01:05 -0000
@@ -1264,6 +1264,12 @@ nsTreeBodyFrame::GetRowAt(PRInt32 aX, PR
   return row;
 }
 
+PRBool
+nsTreeBodyFrame::IsInVisibleSpace(PRInt32 aX, PRInt32 aY)
+{
+  return (aY <= (mRowHeight * mPageLength));
+}
+
 void
 nsTreeBodyFrame::CheckTextForBidi(nsAutoString& aText)
 {
@@ -2604,7 +2610,9 @@ nsTreeBodyFrame::HandleEvent(nsPresConte
         mSlots->mTimer = nsnull;
       }
 
-      if (mSlots->mDropRow >= 0) {
+      if ((mSlots->mDropRow >= 0) ||
+          ((mSlots->mDropRow == -1) &&
+           (mSlots->mDropOrient == nsITreeView::DROP_AFTER))) {
         if (!mSlots->mTimer && mSlots->mDropOrient == nsITreeView::DROP_ON) {
           // Either there wasn't a timer running or it was just killed above.
           // If over a folder, start up a timer to open the folder.
@@ -4231,6 +4239,14 @@ nsTreeBodyFrame::ComputeDropPosition(nsG
         *aOrient = nsITreeView::DROP_AFTER;
     }
   }
+  else if (IsInVisibleSpace(xTwips, yTwips)) {
+    // Drop after last row.  Return with drop after even if there are no rows.
+    if (mRowCount > 0)
+      *aRow = mRowCount - 1;
+    else
+      *aRow = -1;
+    *aOrient = nsITreeView::DROP_AFTER;
+  }
 
   if (CanAutoScroll(*aRow)) {
     // Get the max value from the look and feel service.
Index: layout/xul/base/src/tree/src/nsTreeBodyFrame.h
===================================================================
RCS file: /cvsroot/mozilla/layout/xul/base/src/tree/src/nsTreeBodyFrame.h,v
retrieving revision 1.121
diff -u -p -r1.121 nsTreeBodyFrame.h
--- layout/xul/base/src/tree/src/nsTreeBodyFrame.h	23 Aug 2007 15:57:55 -0000	1.121
+++ layout/xul/base/src/tree/src/nsTreeBodyFrame.h	26 Oct 2007 03:01:05 -0000
@@ -241,6 +241,11 @@ protected:
   // coordinate system of this frame.
   PRInt32 GetRowAt(nscoord aX, nscoord aY);
 
+  // An internal visible hit test.  Returns true if coordinates are in visible
+  // space of frame, even if they're not in a tree row.  aX and aY are expected
+  // to be in app units in the coordinate system of this frame.
+  PRBool IsInVisibleSpace(nscoord aX, nscoord aY);
+
   // Check for bidi characters in the text, and if there are any, ensure
   // that the prescontext is in bidi mode.
   void CheckTextForBidi(nsAutoString& aText);
