Index: config/mkdepend/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/config/mkdepend/Makefile.in,v
retrieving revision 1.16
diff -u -8 -r1.16 Makefile.in
--- config/mkdepend/Makefile.in	21 Feb 2007 20:06:33 -0000	1.16
+++ config/mkdepend/Makefile.in	12 Jun 2008 09:52:56 -0000
@@ -35,16 +35,19 @@
 #
 # ***** END LICENSE BLOCK *****
 
 DEPTH		= ../..
 topsrcdir	= @top_srcdir@
 srcdir		= @srcdir@
 VPATH		= @srcdir@
 
+DISABLE_MOZ_MEMORY = 1
+USE_STATIC_LIBS = 1
+
 include $(DEPTH)/config/autoconf.mk
 
 # undefine (as best we can, thanks gmake!) so we don't need build_number
 MOZILLA_OFFICIAL =
 BUILD_OFFICIAL   =
 MODULE		 = mkdepend
 HOST_PROGRAM	 = mkdepend$(BIN_SUFFIX)
 ifdef GNU_CC
Index: modules/libmar/tool/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/modules/libmar/tool/Makefile.in,v
retrieving revision 1.7
diff -u -8 -r1.7 Makefile.in
--- modules/libmar/tool/Makefile.in	16 Apr 2006 16:29:28 -0000	1.7
+++ modules/libmar/tool/Makefile.in	12 Jun 2008 09:52:56 -0000
@@ -36,16 +36,19 @@
 #
 # ***** END LICENSE BLOCK *****
 
 DEPTH		= ../../..
 topsrcdir	= @top_srcdir@
 srcdir		= @srcdir@
 VPATH		= @srcdir@
 
+DISABLE_MOZ_MEMORY = 1
+USE_STATIC_LIBS = 1
+
 include $(DEPTH)/config/autoconf.mk
 
 MODULE		= mar
 ifeq ($(OS_ARCH),WINNT)
 USE_STATIC_LIBS = 1
 endif
 
 # The mar executable is output into dist/host/bin since it is something that
Index: toolkit/crashreporter/client/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/toolkit/crashreporter/client/Makefile.in,v
retrieving revision 1.19
diff -u -8 -r1.19 Makefile.in
--- toolkit/crashreporter/client/Makefile.in	5 May 2008 21:08:15 -0000	1.19
+++ toolkit/crashreporter/client/Makefile.in	12 Jun 2008 09:52:56 -0000
@@ -37,16 +37,19 @@
 # ***** END LICENSE BLOCK *****
 
 
 DEPTH		= ../../..
 topsrcdir	= @top_srcdir@
 srcdir		= @srcdir@
 VPATH		= @srcdir@
 
+DISABLE_MOZ_MEMORY = 1
+USE_STATIC_LIBS = 1
+
 include $(DEPTH)/config/autoconf.mk
 
 PROGRAM = crashreporter$(BIN_SUFFIX)
 DIST_PROGRAM = crashreporter$(BIN_SUFFIX)
 REQUIRES	= \
 		$(NULL)
 
 LOCAL_INCLUDES = -I$(srcdir)/../google-breakpad/src
Index: xpcom/typelib/xpidl/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/xpcom/typelib/xpidl/Makefile.in,v
retrieving revision 1.44
diff -u -8 -r1.44 Makefile.in
--- xpcom/typelib/xpidl/Makefile.in	23 Feb 2008 00:22:48 -0000	1.44
+++ xpcom/typelib/xpidl/Makefile.in	12 Jun 2008 09:52:56 -0000
@@ -35,16 +35,19 @@
 #
 # ***** END LICENSE BLOCK *****
 
 DEPTH		= ../../..
 topsrcdir	= @top_srcdir@
 srcdir		= @srcdir@
 VPATH		= @srcdir@
 
+DISABLE_MOZ_MEMORY = 1
+USE_STATIC_LIBS = 1
+
 include $(DEPTH)/config/autoconf.mk
 
 MODULE		= xpcom
 
 ifndef WINCE
 # Sadly, the code here is too smart for the WinCE compiler's brain
 PROGRAM		= xpidl$(BIN_SUFFIX)
 SDK_BINARY  = $(PROGRAM)
Index: xpcom/typelib/xpt/tools/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/xpcom/typelib/xpt/tools/Makefile.in,v
retrieving revision 1.32
diff -u -8 -r1.32 Makefile.in
--- xpcom/typelib/xpt/tools/Makefile.in	17 Mar 2008 17:39:01 -0000	1.32
+++ xpcom/typelib/xpt/tools/Makefile.in	12 Jun 2008 09:52:56 -0000
@@ -35,16 +35,18 @@
 #
 # ***** END LICENSE BLOCK *****
 
 DEPTH		= ../../../..
 topsrcdir	= @top_srcdir@
 srcdir		= @srcdir@
 VPATH		= @srcdir@
 
+DISABLE_MOZ_MEMORY = 1 
+
 include $(DEPTH)/config/autoconf.mk
 
 MODULE		= xpcom
 INTERNAL_TOOLS	= 1
 
 SIMPLE_PROGRAMS	= xpt_dump$(BIN_SUFFIX) xpt_link$(BIN_SUFFIX)
 
 CSRCS		= xpt_dump.c xpt_link.c
