Index: layout/xul/base/src/nsProgressMeterFrame.cpp
===================================================================
RCS file: /cvsroot/mozilla/layout/xul/base/src/nsProgressMeterFrame.cpp,v
retrieving revision 1.64
diff -p -u -r1.64 nsProgressMeterFrame.cpp
--- layout/xul/base/src/nsProgressMeterFrame.cpp	1 Dec 2007 07:22:44 -0000	1.64
+++ layout/xul/base/src/nsProgressMeterFrame.cpp	2 Jul 2008 17:43:56 -0000
@@ -129,15 +129,27 @@ nsProgressMeterFrame::AttributeChanged(P
     nsCOMPtr<nsIContent> remainderContent = remainderChild->GetContent();
     if (!remainderContent) return NS_OK;
 
-    nsAutoString value;
+    nsAutoString value, maxValue;
     mContent->GetAttr(kNameSpaceID_None, nsGkAtoms::value, value);
+    mContent->GetAttr(kNameSpaceID_None, nsGkAtoms::max, maxValue);
 
     PRInt32 error;
     PRInt32 flex = value.ToInteger(&error);
-    if (flex < 0) flex = 0;
-    if (flex > 100) flex = 100;
+    PRInt32 maxFlex = maxValue.ToInteger(&error);
+    if (maxValue.IsEmpty()) {
+      maxFlex = 100;
+    }
+    if (maxFlex < 1) {
+      maxFlex = 1;
+    }
+    if (flex < 0) {
+      flex = 0;
+    }
+    if (flex > maxFlex) {
+      flex = maxFlex;
+    }
 
-    PRInt32 remainder = 100 - flex;
+    PRInt32 remainder = maxFlex - flex;
 
     nsAutoString leftFlex, rightFlex;
     leftFlex.AppendInt(flex);
Index: toolkit/content/tests/reftests/bug-442419-progressmeter-max-ref.xul
===================================================================
RCS file: toolkit/content/tests/reftests/bug-442419-progressmeter-max-ref.xul
diff -N toolkit/content/tests/reftests/bug-442419-progressmeter-max-ref.xul
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ toolkit/content/tests/reftests/bug-442419-progressmeter-max-ref.xul	2 Jul 2008 17:43:56 -0000
@@ -0,0 +1,7 @@
+<?xml version="1.0"?>
+<?xml-stylesheet href="chrome://global/skin/global.css" type="text/css"?>
+
+<window xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul">
+  <progressmeter value="50"/> <!-- default is max = 100 -->
+</window>
+
Index: toolkit/content/tests/reftests/bug-442419-progressmeter-max.xul
===================================================================
RCS file: toolkit/content/tests/reftests/bug-442419-progressmeter-max.xul
diff -N toolkit/content/tests/reftests/bug-442419-progressmeter-max.xul
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ toolkit/content/tests/reftests/bug-442419-progressmeter-max.xul	2 Jul 2008 17:43:56 -0000
@@ -0,0 +1,7 @@
+<?xml version="1.0"?>
+<?xml-stylesheet href="chrome://global/skin/global.css" type="text/css"?>
+
+<window xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul">
+  <progressmeter  max="198" value="99"/> <!-- 50% -->
+</window>
+
Index: toolkit/content/tests/reftests/reftest.list
===================================================================
RCS file: toolkit/content/tests/reftests/reftest.list
diff -N toolkit/content/tests/reftests/reftest.list
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ toolkit/content/tests/reftests/reftest.list	2 Jul 2008 17:43:56 -0000
@@ -0,0 +1 @@
+== bug-442419-progressmeter-max.xul bug-442419-progressmeter-max-ref.xul
Index: toolkit/content/widgets/progressmeter.xml
===================================================================
RCS file: /cvsroot/mozilla/toolkit/content/widgets/progressmeter.xml,v
retrieving revision 1.12
diff -p -u -r1.12 progressmeter.xml
--- toolkit/content/widgets/progressmeter.xml	7 Apr 2008 22:56:43 -0000	1.12
+++ toolkit/content/widgets/progressmeter.xml	2 Jul 2008 17:43:56 -0000
@@ -22,16 +22,17 @@
       <property name="value" onget="return this.getAttribute('value') || '0';">
         <setter><![CDATA[
           var p = Math.round(val);
+          var max = Math.round(this.max);
           if (p < 0)
             p = 0;
-          else if (p > 100)
-            p = 100;
+          else if (p > max)
+            p = max;
           var c = this.value; 
           if (p != c) {
             var delta = p - c;
             if (delta < 0)
               delta = -delta;
-            if (delta > 3 || p == 0 || p == 100) {
+            if (delta > 3 || p == 0 || p == max) {
               this.setAttribute("value", p);
               // Fire DOM event so that accessible value change events occur
               var event = document.createEvent('Events');
@@ -43,6 +44,9 @@
           return val;
         ]]></setter>
       </property>
+      <property name="max"
+                onget="return this.getAttribute('max') || '100';"
+                onset="this.setAttribute('max', Math.max(val, 1));"/>
 
       <property name="accessibleType" readonly="true">
         <getter>
Index: widget/src/cocoa/nsNativeThemeCocoa.h
===================================================================
RCS file: /cvsroot/mozilla/widget/src/cocoa/nsNativeThemeCocoa.h,v
retrieving revision 1.20
diff -p -u -r1.20 nsNativeThemeCocoa.h
--- widget/src/cocoa/nsNativeThemeCocoa.h	9 May 2008 18:12:39 -0000	1.20
+++ widget/src/cocoa/nsNativeThemeCocoa.h	2 Jul 2008 17:43:56 -0000
@@ -100,7 +100,7 @@ protected:  
                   PRInt32 inState);
   void DrawProgress(CGContextRef context, const HIRect& inBoxRect,
                     PRBool inIsIndeterminate, PRBool inIsHorizontal,
-                    PRInt32 inValue);
+                    PRInt32 inValue, PRInt32 inMaxValue);
   void DrawTab (CGContextRef context, const HIRect& inBoxRect,
                 PRBool inIsDisabled, PRBool inIsFrontmost, 
                 PRBool inIsHorizontal, PRBool inTabBottom,
Index: widget/src/cocoa/nsNativeThemeCocoa.mm
===================================================================
RCS file: /cvsroot/mozilla/widget/src/cocoa/nsNativeThemeCocoa.mm,v
retrieving revision 1.93
diff -p -u -r1.93 nsNativeThemeCocoa.mm
--- widget/src/cocoa/nsNativeThemeCocoa.mm	9 May 2008 18:12:39 -0000	1.93
+++ widget/src/cocoa/nsNativeThemeCocoa.mm	2 Jul 2008 17:43:56 -0000
@@ -672,8 +672,11 @@ nsNativeThemeCocoa::DrawFrame(CGContextR
 
 void
 nsNativeThemeCocoa::DrawProgress(CGContextRef cgContext,
-                                 const HIRect& inBoxRect, PRBool inIsIndeterminate, 
-                                 PRBool inIsHorizontal, PRInt32 inValue)
+                                 const HIRect& inBoxRect, 
+                                 PRBool inIsIndeterminate, 
+                                 PRBool inIsHorizontal, 
+                                 PRInt32 inValue, 
+                                 PRInt32 inMaxValue)
 {
   NS_OBJC_BEGIN_TRY_ABORT_BLOCK;
 
@@ -686,7 +689,7 @@ nsNativeThemeCocoa::DrawProgress(CGConte
   tdi.kind = inIsIndeterminate ? kThemeMediumIndeterminateBar: kThemeMediumProgressBar;
   tdi.bounds = inBoxRect;
   tdi.min = 0;
-  tdi.max = 100;
+  tdi.max = inMaxValue;
   tdi.value = inValue;
   tdi.attributes = inIsHorizontal ? kThemeTrackHorizontal : 0;
   tdi.enableState = kThemeTrackActive;
@@ -1195,12 +1198,12 @@ nsNativeThemeCocoa::DrawWidgetBackground
       
     case NS_THEME_PROGRESSBAR:
       DrawProgress(cgContext, macRect, IsIndeterminateProgress(aFrame),
-                   PR_TRUE, GetProgressValue(aFrame));
+                   PR_TRUE, GetProgressValue(aFrame), GetProgressMaxValue(aFrame));
       break;
 
     case NS_THEME_PROGRESSBAR_VERTICAL:
       DrawProgress(cgContext, macRect, IsIndeterminateProgress(aFrame),
-                   PR_FALSE, GetProgressValue(aFrame));
+                   PR_FALSE, GetProgressValue(aFrame), GetProgressMaxValue(aFrame));
       break;
 
     case NS_THEME_PROGRESSBAR_CHUNK:
Index: widget/src/xpwidgets/nsNativeTheme.h
===================================================================
RCS file: /cvsroot/mozilla/widget/src/xpwidgets/nsNativeTheme.h,v
retrieving revision 1.28
diff -p -u -r1.28 nsNativeTheme.h
--- widget/src/xpwidgets/nsNativeTheme.h	27 Mar 2008 05:36:21 -0000	1.28
+++ widget/src/xpwidgets/nsNativeTheme.h	2 Jul 2008 17:43:56 -0000
@@ -133,6 +133,10 @@ class nsNativeTheme
   PRInt32 GetProgressValue(nsIFrame* aFrame) {
     return CheckIntAttr(aFrame, nsWidgetAtoms::value, 0);
   }
+  
+  PRInt32 GetProgressMaxValue(nsIFrame* aFrame) {
+    return PR_MAX(CheckIntAttr(aFrame, nsWidgetAtoms::max, 100), 1);
+  }
 
   // textfield:
   PRBool IsReadOnly(nsIFrame* aFrame) {
Index: widget/src/xpwidgets/nsWidgetAtomList.h
===================================================================
RCS file: /cvsroot/mozilla/widget/src/xpwidgets/nsWidgetAtomList.h,v
retrieving revision 1.26
diff -p -u -r1.26 nsWidgetAtomList.h
--- widget/src/xpwidgets/nsWidgetAtomList.h	7 Apr 2008 22:56:44 -0000	1.26
+++ widget/src/xpwidgets/nsWidgetAtomList.h	2 Jul 2008 17:43:56 -0000
@@ -85,6 +85,7 @@ WIDGET_ATOM(input, "input")
 WIDGET_ATOM(key, "key") // The key element / attribute
 WIDGET_ATOM(label, "label")
 WIDGET_ATOM(lasttab, "last-tab")
+WIDGET_ATOM(max, "max")
 WIDGET_ATOM(maxpos, "maxpos")
 WIDGET_ATOM(menu, "menu") // Represents an XP menu
 WIDGET_ATOM(menuitem, "menuitem") // Represents an XP menu item
