Index: jpeg/jdapimin.c
===================================================================
RCS file: /cvsroot/mozilla/jpeg/jdapimin.c,v
retrieving revision 3.9
diff -p -U 8 -r3.9 jdapimin.c
--- jpeg/jdapimin.c	25 Jan 2008 07:15:32 -0000	3.9
+++ jpeg/jdapimin.c	26 Jan 2008 00:58:26 -0000
@@ -16,17 +16,37 @@
  * case.
  */
 
 #define JPEG_INTERNALS
 #include "jinclude.h"
 #include "jpeglib.h"
 
 #ifdef HAVE_MMX_INTEL_MNEMONICS
+#if _MSC_VER >= 0x800
 #include "intrin.h"
+#else
+/* no __cpuid intrinsic, use a slower replacement */
+void __cpuid( int CPUInfo[4], int InfoType )
+{
+  int my_eax, my_ebx, my_ecx, my_edx;
+  __asm {
+    mov eax, InfoType;
+    cpuid
+    mov my_eax, eax
+    mov my_ebx, ebx
+    mov my_ecx, ecx
+    mov my_edx, edx
+  }
+  CPUInfo[0] = my_eax;
+  CPUInfo[1] = my_ebx;
+  CPUInfo[2] = my_ecx;
+  CPUInfo[3] = my_edx;
+}
+#endif /* _MSC_VER >= 0x800 */
 int MMXAvailable;
 static int mmxsupport();
 #endif
 
 #ifdef HAVE_SSE2_INTRINSICS
 int SSE2Available = 0;
 #ifdef HAVE_SSE2_INTEL_MNEMONICS
 static int sse2support();
