Index: widget/src/cocoa/nsCocoaWindow.h
===================================================================
RCS file: /cvsroot/mozilla/widget/src/cocoa/nsCocoaWindow.h,v
retrieving revision 1.43
diff -u -8 -p -r1.43 nsCocoaWindow.h
--- widget/src/cocoa/nsCocoaWindow.h	30 Oct 2007 04:03:43 -0000	1.43
+++ widget/src/cocoa/nsCocoaWindow.h	7 Dec 2007 20:12:48 -0000
@@ -94,17 +94,21 @@ class nsChildView;
 
 
 @interface BorderlessWindow : NSWindow
 {
 }
 
 - (BOOL)canBecomeKeyWindow;
 - (BOOL)canBecomeMainWindow;
+@end
+
 
+@interface SongbirdWindow : NSWindow 
+{}
 @end
 
 
 @interface WindowDelegate : NSObject
 {
   nsCocoaWindow* mGeckoWindow; // [WEAK] (we are owned by the window)
 }
 - (id)initWithGeckoWindow:(nsCocoaWindow*)geckoWind;
Index: widget/src/cocoa/nsCocoaWindow.mm
===================================================================
RCS file: /cvsroot/mozilla/widget/src/cocoa/nsCocoaWindow.mm,v
retrieving revision 1.117
diff -u -8 -p -r1.117 nsCocoaWindow.mm
--- widget/src/cocoa/nsCocoaWindow.mm	17 Nov 2007 12:49:04 -0000	1.117
+++ widget/src/cocoa/nsCocoaWindow.mm	7 Dec 2007 20:12:48 -0000
@@ -325,20 +325,34 @@ nsresult nsCocoaWindow::StandardCreate(n
     // Note that we need to check the window type because we mark sheets sheets as 
     // having titlebars.
     if (mWindowType == eWindowType_toplevel &&
         (features & NSTitledWindowMask))
       windowClass = [ToolbarWindow class];
     // If we're a popup window we need to use the PopupWindow class.
     else if (mWindowType == eWindowType_popup)
       windowClass = [PopupWindow class];
-    // If we're a non-popup borderless window we need to use the
-    // BorderlessWindow class.
-    else if (features == NSBorderlessWindowMask)
-      windowClass = [BorderlessWindow class];
+      
+    // XXXmatt: If we use a borderless window the OS makes
+    // assumptions about what the window should be able to do.
+    // What we really want is a normal window in which
+    // the XUL takes responsibility for drawing the titlebar
+    // and resizer.
+    else if (features == NSBorderlessWindowMask) {
+      windowClass = [SongbirdWindow class];
+      features |= NSClosableWindowMask;
+      features |= NSTitledWindowMask;
+      features |= NSMiniaturizableWindowMask;
+      
+      // If the window was opened with resizable=yes, then 
+      // enable to OS resizer.
+      if (aInitData->mBorderStyle & eBorderStyle_resizeh) {
+        features |= NSResizableWindowMask;
+      }
+    }
 
     // Create the window
     mWindow = [[windowClass alloc] initWithContentRect:rect styleMask:features 
                                    backing:NSBackingStoreBuffered defer:NO];
     
     if (mWindowType == eWindowType_popup) {
       [mWindow setAlphaValue:0.95];
       [mWindow setLevel:NSPopUpMenuWindowLevel];
@@ -357,17 +371,21 @@ nsresult nsCocoaWindow::StandardCreate(n
         [mWindow setContentView:newContentView];
       }
     }
     else if (mWindowType == eWindowType_invisible) {
       [mWindow setLevel:kCGDesktopWindowLevelKey];
     }
 
     [mWindow setBackgroundColor:[NSColor whiteColor]];
-    [mWindow setContentMinSize:NSMakeSize(60, 60)];
+    
+    // XXXmatt: Modifying the arbitrary min size so that it 
+    // doesn't interfere with the Songbird miniplayer.
+    [mWindow setContentMinSize:NSMakeSize(60, 20)];
+    
     [mWindow setReleasedWhenClosed:NO];
 
     // setup our notification delegate. Note that setDelegate: does NOT retain.
     mDelegate = [[WindowDelegate alloc] initWithGeckoWindow:this];
     [mWindow setDelegate:mDelegate];
     
     mWindowMadeHere = PR_TRUE;
   }
@@ -1822,8 +1840,171 @@ void patternDraw(void* aInfo, CGContextR
 {
   NSWindow *nativeWindow = [self retain];
   BOOL retval = [super performKeyEquivalent:theEvent];
   [nativeWindow release];
   return retval;
 }
 
 @end
+
+
+
+
+/*************************************************************************
+ * BEGIN SONGBIRD HACK
+ *
+ * If you want to draw your own window decorations like iPhoto or
+ * Aperature, you have to use the unsupported NSThemeFrame view.
+ *
+ * See http://andymatuschak.org/articles/2006/01/11/ for gory details.
+ *
+ * In Songbird we like to draw all window decorations using XUL, so we
+ * subclass NSThemeFrame and make the content area 100% of the window. 
+ * BorderlessWindow would work, but causes the OS to make assumptions
+ * about things like resizing and zooming.
+ *
+ *************************************************************************/
+
+
+
+@class NSDocumentDragButton, NSButton, NSCell, NSImage, NSString;
+
+@interface NSFrameView : NSView
+{
+  unsigned int styleMask;
+  NSString *_title;
+  NSCell *titleCell;
+  NSButton *closeButton;
+  NSButton *zoomButton;
+  NSButton *minimizeButton;
+  char resizeByIncrement;
+  char frameNeedsDisplay;
+  unsigned char tabViewCount;
+  struct _NSSize resizeParameter;
+  int shadowState; 
+}
+
+- (id)initWithFrame:(struct _NSRect)fp8 styleMask:(unsigned int)fp24 owner:(id)fp28;
+- (id)initWithFrame:(struct _NSRect)fp8;
+
+@end
+
+
+
+
+@interface NSTitledFrame : NSFrameView
+{
+  int resizeFlags;
+  id fileButton;    
+  struct _NSSize titleCellSize;
+}
+
+- (id)initWithFrame:(struct _NSRect)fp8 styleMask:(unsigned int)fp24 owner:(id)fp28;
+- (float)_titlebarHeight;
++ (float)_titlebarHeight:(unsigned int)fp8;
+- (int)_numberOfTitlebarLines;
++ (float)_windowTitlebarTitleMinHeight:(unsigned int)fp8;
+- (float)_windowTitlebarTitleMinHeight;
+
+@end
+
+
+
+
+@interface NSThemeFrame : NSTitledFrame
+{
+  NSButton *toolbarButton;
+  int toolbarVisibleStatus;
+  NSImage *showToolbarTransitionImage;
+  struct _NSSize showToolbarPreWindowSize;
+  NSButton *modeButton;
+  int leftGroupTrackingTagNum;
+  int rightGroupTrackingTagNum;
+  char mouseInsideLeftGroup;
+  char mouseInsideRightGroup;
+  int widgetState;
+  NSString *displayName; 
+}
+
+- (id)initWithFrame:(struct _NSRect)fp8 styleMask:(unsigned int)fp24 owner:(id)fp28;
+- (float)_titlebarHeight;
++ (float)_titlebarHeight:(unsigned int)fp8;
+- (float)_windowTitlebarTitleMinHeight;
++ (float)_windowTitlebarTitleMinHeight:(unsigned int)fp8;
+
+@end
+
+
+
+
+// A ThemeFrame with no titlebar or resize indicator,
+
+@interface SongbirdThemeFrame : NSThemeFrame {}
+@end
+
+@implementation SongbirdThemeFrame
+
+- initWithFrame:(NSRect)frame styleMask:(int)sm owner:owner
+{
+  self = [super initWithFrame:frame styleMask:sm owner:owner];
+
+  if (self) {
+    NSWindow *window = [self window];
+    
+    // Turn off the standard window controls
+    [[window standardWindowButton:NSWindowMiniaturizeButton] setHidden:YES];
+    [[window standardWindowButton:NSWindowZoomButton] setHidden:YES];
+    [[window standardWindowButton:NSWindowCloseButton] setHidden:YES];
+
+    // Let the window paint its own resize indicator.
+    // Note that we are suppressing the graphic, and
+    // the resizer may still be functional.
+    [window setShowsResizeIndicator:NO];
+
+    [window setHasShadow:YES];
+  }
+  return self;
+}
+
+// Crush the titlebar, since the window will paint its own
+
++ (float)_windowTitlebarTitleMinHeight:(unsigned int)fp8 
+{
+  return 0;
+}
+
+- (int)titlebarHeight
+{
+  return 0;
+}
+
+- (float)_titlebarHeight
+{
+  return 0.0;
+}
+
+- (int)_numberOfTitlebarLines 
+{
+  return 0;
+}
+
+@end
+
+
+
+
+// Create a window that uses our special frame
+
+@interface NSWindow (Private)
++ (Class)frameViewClassForStyleMask:(unsigned int)mask;
+@end
+
+@implementation SongbirdWindow
++ (Class)frameViewClassForStyleMask:(unsigned int)styleMask
+{
+  return [SongbirdThemeFrame class];
+}
+@end
+
+/*************************************************************************
+ * END SONGBIRD HACK
+ ************************************************************************/
