Index: layout/xul/base/src/nsMenuPopupFrame.cpp
===================================================================
RCS file: /cvsroot/mozilla/layout/xul/base/src/nsMenuPopupFrame.cpp,v
retrieving revision 1.338
diff -u -8 -r1.338 nsMenuPopupFrame.cpp
--- layout/xul/base/src/nsMenuPopupFrame.cpp	8 Mar 2008 11:32:26 -0000	1.338
+++ layout/xul/base/src/nsMenuPopupFrame.cpp	30 Apr 2008 19:07:30 -0000
@@ -1108,18 +1108,17 @@
          * - - - - - - - - - - screenBottomTwips (bottom of the screen)
          *   |    \ 
          *   |     }  moveDistY
          *   |    / 
          *   +----  screenViewLocY + mRect.height
          */
 
         if ( (screenViewLocY + mRect.height) > screenBottomTwips ) {
-          // XXX Bug 84121 comment 48 says the next line has to use screenHeightTwips, why not screenBottomTwips?
-          PRInt32 moveDistY = (screenViewLocY + mRect.height) - screenHeightTwips;
+          PRInt32 moveDistY = (screenViewLocY + mRect.height) - screenBottomTwips;
           if ( screenViewLocY - moveDistY < screenTopTwips )
             moveDistY = screenViewLocY - screenTopTwips;          
           screenViewLocY -= moveDistY;
           ypos -= moveDistY; 
         } 
       }
       
       // Resize it to fit on the screen. By this point, we've given the popup as much
