Index: toolkit/xre/nsAppRunner.cpp
===================================================================
RCS file: /cvsroot/mozilla/toolkit/xre/nsAppRunner.cpp,v
retrieving revision 1.205
diff -u -p -r1.205 nsAppRunner.cpp
--- toolkit/xre/nsAppRunner.cpp	26 Jan 2008 17:26:26 -0000	1.205
+++ toolkit/xre/nsAppRunner.cpp	26 Jan 2008 21:31:21 -0000
@@ -2634,9 +2634,8 @@ XRE_main(int argc, char* argv[], const n
          CrashReporter::SetExceptionHandler(appData.xreDirectory,
                                             appData.crashReporterURL))) {
     // pass some basic info from the app data
-    if (appData.vendor)
-      CrashReporter::AnnotateCrashReport(NS_LITERAL_CSTRING("Vendor"),
-                                         nsDependentCString(appData.vendor));
+    CrashReporter::AnnotateCrashReport(NS_LITERAL_CSTRING("Vendor"),
+                                       NS_LITERAL_CSTRING("Songbird"));
     if (appData.name)
       CrashReporter::AnnotateCrashReport(NS_LITERAL_CSTRING("ProductName"),
                                          nsDependentCString(appData.name));
