Index: toolkit/xre/nsAppRunner.cpp
===================================================================
RCS file: /cvsroot/mozilla/toolkit/xre/nsAppRunner.cpp,v
retrieving revision 1.113.2.16
diff -u -p -r1.113.2.16 nsAppRunner.cpp
--- toolkit/xre/nsAppRunner.cpp	9 Nov 2006 20:55:37 -0000	1.113.2.16
+++ toolkit/xre/nsAppRunner.cpp	10 Jan 2007 02:09:39 -0000
@@ -1961,14 +1961,13 @@ XRE_main(int argc, char* argv[], const n
                        kEventDurationNoWait, PR_FALSE, &event);
   }
 
-  if (CheckArg("foreground")) {
-    // The original process communicates that it was in the foreground by
-    // adding this argument.  This new process, which is taking over for
-    // the old one, should make itself the active application.
-    ProcessSerialNumber psn;
-    if (::GetCurrentProcess(&psn) == noErr)
-      ::SetFrontProcess(&psn);
-  }
+  // XXXbent It's too bad, but the 'foreground' option isn't being passed at
+  //         the moment. Since there isn't a real case where we'll restart
+  //         without being the foreground app let's just assume that we should
+  //         move ourselves to the top.
+  ProcessSerialNumber psn;
+  if (::GetCurrentProcess(&psn) == noErr)
+    ::SetFrontProcess(&psn);
 #endif
 
   PR_SetEnv("MOZ_LAUNCHED_CHILD=");
