Index: content/base/src/nsDocument.cpp
===================================================================
RCS file: /cvsroot/mozilla/content/base/src/nsDocument.cpp,v
retrieving revision 3.566.2.26
diff -u -p -r3.566.2.26 nsDocument.cpp
--- content/base/src/nsDocument.cpp	3 Aug 2006 00:47:50 -0000	3.566.2.26
+++ content/base/src/nsDocument.cpp	21 Sep 2006 16:45:33 -0000
@@ -3168,6 +3168,8 @@ nsDocument::GetBoxObjectFor(nsIDOMElemen
 {
   nsCOMPtr<nsIContent> content(do_QueryInterface(aElement));
   NS_ENSURE_TRUE(content, NS_ERROR_UNEXPECTED);
+
+  // This check must match the one in nsXULElement.cpp::GetBoxObject()
   NS_ENSURE_TRUE(content->GetOwnerDoc() == this,
                  NS_ERROR_DOM_WRONG_DOCUMENT_ERR);
   
Index: content/xul/content/src/nsXULElement.cpp
===================================================================
RCS file: /cvsroot/mozilla/content/xul/content/src/nsXULElement.cpp,v
retrieving revision 1.578.2.17
diff -u -p -r1.578.2.17 nsXULElement.cpp
--- content/xul/content/src/nsXULElement.cpp	3 Aug 2006 00:47:51 -0000	1.578.2.17
+++ content/xul/content/src/nsXULElement.cpp	21 Sep 2006 16:45:34 -0000
@@ -2589,7 +2589,9 @@ nsXULElement::GetBoxObject(nsIBoxObject*
   *aResult = nsnull;
 
   // XXX sXBL/XBL2 issue! Owner or current document?
-  nsCOMPtr<nsIDOMNSDocument> nsDoc(do_QueryInterface(GetCurrentDoc()));
+  // Be sure to get the same document as the NS_ENSURE_TRUE uses in
+  // nsDocument.cpp::GetBoxObjectFor().
+  nsCOMPtr<nsIDOMNSDocument> nsDoc(do_QueryInterface(GetOwnerDoc()));
   NS_ENSURE_TRUE(nsDoc, NS_ERROR_FAILURE);
   return nsDoc->GetBoxObjectFor(this, aResult);
 }
