Index: js/src/liveconnect/nsCLiveconnectFactory.cpp
===================================================================
RCS file: /cvsroot/mozilla/js/src/liveconnect/nsCLiveconnectFactory.cpp,v
retrieving revision 1.18
diff -u -d -p -r1.18 nsCLiveconnectFactory.cpp
--- js/src/liveconnect/nsCLiveconnectFactory.cpp	1 Nov 2004 18:50:18 -0000	1.18
+++ js/src/liveconnect/nsCLiveconnectFactory.cpp	8 Sep 2006 00:23:30 -0000
@@ -58,36 +58,6 @@ static NS_DEFINE_IID(kISupportsIID,    N
 static NS_DEFINE_IID(kIFactoryIID,     NS_IFACTORY_IID);
 static NS_DEFINE_CID(kCLiveconnectCID, NS_CLIVECONNECT_CID);
 
-/*+++++++++++++++++++++++++++++++++++++++++++++++++
- * NSGetFactory:
- * Provides entry point to liveconnect dll.
- +++++++++++++++++++++++++++++++++++++++++++++++++*/
-
-extern "C" NS_EXPORT nsresult
-NSGetFactory(nsISupports* serviceMgr,
-             const nsCID &aClass,
-             const char *aClassName,
-             const char *aContractID,
-             nsIFactory **aFactory)
-{
-    
-    if (!aClass.Equals(kCLiveconnectCID)) {
-        return NS_ERROR_FACTORY_NOT_LOADED;     // XXX right error?
-    }
-    nsCLiveconnectFactory* factory = new nsCLiveconnectFactory();
-    if (factory == NULL)
-        return NS_ERROR_OUT_OF_MEMORY;
-    factory->AddRef();
-    *aFactory = factory;
-    return NS_OK;
-}
-
-extern "C" NS_EXPORT PRBool
-NSCanUnload(nsISupports* serviceMgr)
-{
-    return PR_FALSE;
-}
-
 extern "C" NS_EXPORT nsresult
 JSJ_RegisterLiveConnectFactory()
 {
