Index: toolkit/xre/nsUpdateDriver.cpp
===================================================================
RCS file: /cvsroot/mozilla/toolkit/xre/nsUpdateDriver.cpp,v
retrieving revision 1.11.2.7
diff -u -8 -p -r1.11.2.7 nsUpdateDriver.cpp
--- toolkit/xre/nsUpdateDriver.cpp	20 Jul 2006 19:48:45 -0000	1.11.2.7
+++ toolkit/xre/nsUpdateDriver.cpp	23 Aug 2006 20:27:30 -0000
@@ -48,16 +48,17 @@
 #include "nsCOMPtr.h"
 #include "nsCOMArray.h"
 #include "nsString.h"
 #include "nsPrintfCString.h"
 #include "prproces.h"
 #include "prlog.h"
 
 #ifdef XP_MACOSX
+#include "nsILocalFileMac.h"
 #include "nsCommandLineServiceMac.h"
 #endif
 
 #if defined(XP_WIN)
 # include <direct.h>
 # include <process.h>
 # include <windows.h>
 # define getcwd(path, size) _getcwd(path, size)
@@ -120,16 +121,58 @@ GetCurrentWorkingDir(char *buf, size_t s
     return NS_ERROR_FAILURE;
 #else
   if(!getcwd(buf, size))
     return NS_ERROR_FAILURE;
 #endif
   return NS_OK;
 }
 
+#if defined(MOZ_XULRUNNER) && defined(XP_MACOSX)
+
+// This is a copy of OS X's XRE_GetBinaryPath from nsAppRunner.cpp with the
+// gBinaryPath check removed so that the updater can reload the stub executable
+// instead of xulrunner-bin. See bug 349737.
+static nsresult
+GetXULRunnerStubPath(const char* argv0, nsILocalFile* *aResult)
+{
+  nsresult rv;
+  nsCOMPtr<nsILocalFile> lf;
+
+  NS_NewNativeLocalFile(EmptyCString(), PR_TRUE, getter_AddRefs(lf));
+  nsCOMPtr<nsILocalFileMac> lfm (do_QueryInterface(lf));
+  if (!lfm)
+    return NS_ERROR_FAILURE;
+
+  // Works even if we're not bundled.
+  CFBundleRef appBundle = CFBundleGetMainBundle();
+  if (!appBundle)
+    return NS_ERROR_FAILURE;
+
+  CFURLRef bundleURL = CFBundleCopyExecutableURL(appBundle);
+  if (!bundleURL)
+    return NS_ERROR_FAILURE;
+
+  FSRef fileRef;
+  if (!CFURLGetFSRef(bundleURL, &fileRef)) {
+    CFRelease(bundleURL);
+    return NS_ERROR_FAILURE;
+  }
+
+  rv = lfm->InitWithFSRef(&fileRef);
+  CFRelease(bundleURL);
+
+  if (NS_FAILED(rv))
+    return rv;
+
+  NS_ADDREF(*aResult = lf);
+  return NS_OK;
+}
+#endif /* MOZ_XULRUNNER && XP_MACOSX */
+
 PR_STATIC_CALLBACK(int)
 ScanDirComparator(nsIFile *a, nsIFile *b, void *unused)
 {
   // lexically compare the leaf names of these two files
   nsCAutoString a_name, b_name;
   a->GetNativeLeafName(a_name);
   b->GetNativeLeafName(b_name);
   return Compare(a_name, b_name);
@@ -294,17 +337,25 @@ ApplyUpdate(nsIFile *greDir, nsIFile *up
   if (!CopyUpdaterIntoUpdateDir(greDir, updateDir, updater)) {
     LOG(("failed copying updater\n"));
     return;
   }
 
   // We need to use the value returned from XRE_GetBinaryPath when attempting
   // to restart the running application.
   nsCOMPtr<nsILocalFile> appFile;
+
+#if defined(MOZ_XULRUNNER) && defined(XP_MACOSX)
+  // On OS X we need to pass the location of the xulrunner-stub executable
+  // rather than xulrunner-bin. See bug 349737.
+  GetXULRunnerStubPath(appArgv[0], getter_AddRefs(appFile));
+#else
   XRE_GetBinaryPath(appArgv[0], getter_AddRefs(appFile));
+#endif
+
   if (!appFile)
     return;
   nsCAutoString appFilePath;
   nsresult rv = appFile->GetNativePath(appFilePath);
   if (NS_FAILED(rv))
     return;
   
   nsCAutoString updaterPath;
