Index: xpfe/bootstrap/nsSigHandlers.cpp
===================================================================
RCS file: /cvsroot/mozilla/xpfe/bootstrap/nsSigHandlers.cpp,v
retrieving revision 1.41.18.1
diff -U8 -r1.41.18.1 nsSigHandlers.cpp
--- xpfe/bootstrap/nsSigHandlers.cpp	22 Jun 2006 22:31:13 -0000	1.41.18.1
+++ xpfe/bootstrap/nsSigHandlers.cpp	4 Aug 2006 01:58:56 -0000
@@ -71,17 +71,17 @@
 #ifdef MOZ_WIDGET_PHOTON
 #include <photon/PhProto.h>
 #include <string.h>
 #endif
 
 static char _progname[1024] = "huh?";
 static unsigned int _gdb_sleep_duration = 300;
 
-#if defined(LINUX) && defined(DEBUG) && (defined(__i386) || defined(PPC))
+#if defined(LINUX) && defined(DEBUG) && (defined(__i386) || defined(PPC) || defined(HAVE_64BIT_OS))
 #define CRAWL_STACK_ON_SIGSEGV
 #endif
  
 #ifdef MOZ_WIDGET_PHOTON
 void abnormal_exit_handler(int signum)
 {
   /* Free any shared memory that has been allocated */
   PgShmemCleanup();
