Index: gfx/cairo/libpixman/src/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/gfx/cairo/libpixman/src/Makefile.in,v
retrieving revision 1.12.4.1
diff -u -8 -r1.12.4.1 Makefile.in
--- gfx/cairo/libpixman/src/Makefile.in	27 Sep 2005 23:09:02 -0000	1.12.4.1
+++ gfx/cairo/libpixman/src/Makefile.in	6 Feb 2006 20:54:46 -0000
@@ -44,19 +44,21 @@
 include $(DEPTH)/config/autoconf.mk
 
 MODULE		= libpixman
 LIBRARY_NAME	= mozlibpixman
 LIBXUL_LIBRARY = 1
 
 ifeq ($(OS_ARCH)_$(GNU_CC), WINNT_)
 ifeq ($(MOZ_OPTIMIZE),1)
+ifeq (1200,$(_MSC_VER))
 USE_STATIC_VC71_LIB = 1
 endif
 endif
+endif
 
 CSRCS	= \
 	fbcompose.c \
 	fbedge.c \
 	fbpict.c \
 	fbtrap.c \
 	icblt.c \
 	icbltone.c \
@@ -85,8 +87,9 @@
 include $(topsrcdir)/config/rules.mk
 
 ifeq ($(USE_STATIC_VC71_LIB),1)
 libs:: $(srcdir)/pixman-vc71.lib
 	cp $(srcdir)/pixman-vc71.lib $(LIBRARY)
 	$(INSTALL) $(IFLAGS1) $(LIBRARY) $(DIST)/lib
 endif
 
+
